sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptors/UserAgentInterceptor.java

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.core.http.interceptors;
+
+import androidx.annotation.NonNull;
+
+import com.azure.android.core.util.Configuration;
+import com.azure.android.core.util.ServiceVersion;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/**
+ * Pipeline interceptor that adds "User-Agent" header to a request.
+ * <p>
+ * The format for the "User-Agent" string is outlined in
+ * <a href="https://azure.github.io/azure-sdk/general_azurecore.html#telemetry-policy">Azure Core: Telemetry policy</a>.
+ */
+public class UserAgentInterceptor implements Interceptor {
+    private static final String DEFAULT_USER_AGENT_HEADER = "azsdk-java";
+
+    // From the design guidelines, the default user agent header format is:
+    // azsdk-java-<client_lib>/<sdk_version> <platform_info>
+    private static final String USER_AGENT_FORMAT = DEFAULT_USER_AGENT_HEADER + "-%s/%s (%s)";
+
+    // From the design guidelines, the application id user agent header format is:
+    // AzCopy/10.0.4-Preview azsdk-java-<client_lib>/<sdk_version> <platform_info>
+    private static final String APPLICATION_ID_USER_AGENT_FORMAT = "%s " + USER_AGENT_FORMAT;
+
+    // When the AZURE_TELEMETRY_DISABLED configuration is true remove the <platform_info> portion of the user agent.
+    private static final String DISABLED_TELEMETRY_USER_AGENT_FORMAT = DEFAULT_USER_AGENT_HEADER + "-%s/%s";
+
+    // From the design guidelines, the platform info format is:
+    // <language runtime>; <os name> <os version>
+    private static final String PLATFORM_INFO_FORMAT = "%s; %s %s";
```

Comment:
This probably needs to be a change in our design guidelines. The runtime is always known, and the "OS" name and version probably don't make sense, instead we probably want to get some info about the Android API level / version, the OEM, etc. This also seems like an Adrian discussion.